File: rawDroid/src/main/java/com/anthonymandra/framework/Histogram.java
package com.anthonymandra.framework;

import android.graphics.Bitmap;
import android.graphics.Color;

import java.util.concurrent.RecursiveTask;

public class Histogram extends RecursiveTask<Histogram.ColorBins>
{
	@SuppressWarnings("unused")
	private static final String TAG = Histogram.class.getSimpleName();
	private static final int sThreshold = 10000;
	private static final int COLOR_DEPTH = 256;
	private final int[] mSource;
	private final int mStart;
	private final int mLength;

	public Histogram(final int[] src, final int start, final int length)
	{
		mSource = src;
		mStart = start;
		mLength = length;
	}

	private ColorBins computeDirectly()
	{
		ColorBins bins = new ColorBins();
		for (int index = mStart; index < mStart + mLength; index++)
		{
			bins.red[Color.red(mSource[index])]++;
			bins.green[Color.green(mSource[index])]++;
			bins.blue[Color.blue(mSource[index])]++;
		}
		return bins;
	}

	protected ColorBins compute()
	{
		if (mLength < sThreshold)
		{
			return computeDirectly();
		}

		int split = mLength / 2;
		Histogram left = new Histogram(mSource, mStart, split);
		Histogram right = new Histogram(mSource, split, mLength - split);
		left.fork();
		ColorBins r = right.compute();
		ColorBins l = left.join();
		r.merge(l);
		return r;
	}

	static Histogram createHistogram(final Bitmap bmp)
	{
		int[] pixels = new int[bmp.getWidth() * bmp.getHeight()];
		bmp.getPixels(pixels, 0, bmp.getWidth(), 0, 0, bmp.getWidth(), bmp.getHeight());
		return new Histogram(pixels, 0, pixels.length);
	}

	public static class ColorBins
	{
		public int[] red = new int[COLOR_DEPTH];
		public int[] green = new int[COLOR_DEPTH];
		public int[] blue = new int[COLOR_DEPTH];

		int maxRed;
		int maxGreen;
		int maxBlue;

		void merge(ColorBins b)
		{
			for (int i = 0; i < COLOR_DEPTH; i++)
			{
				red[i] += b.red[i];
				green[i] += b.green[i];
				blue[i] += b.blue[i];

				maxRed = Math.max(red[i], maxRed);
				maxGreen = Math.max(green[i], maxGreen);
				maxBlue = Math.max(blue[i], maxBlue);
			}
		}

		public int getMaxFreq()
		{
			return Math.max(Math.max(maxRed, maxGreen), maxBlue);
		}
	}
}

File: app/src/main/java/com/anthonymandra/framework/Histogram.kt
package com.anthonymandra.framework

import android.graphics.Bitmap
import android.graphics.Color

import java.util.concurrent.RecursiveTask

class Histogram(private val mSource: IntArray, private val mStart: Int, private val mLength: Int) : RecursiveTask<Histogram.ColorBins>() {

    private fun computeDirectly(): ColorBins {
        val bins = ColorBins()
        for (index in mStart until mStart + mLength) {
            bins.red[Color.red(mSource[index])]++
            bins.green[Color.green(mSource[index])]++
            bins.blue[Color.blue(mSource[index])]++
        }
        return bins
    }

    override fun compute(): ColorBins {
        if (mLength < sThreshold) {
            return computeDirectly()
        }

        val split = mLength / 2
        val left = Histogram(mSource, mStart, split)
        val right = Histogram(mSource, split, mLength - split)
        left.fork()
        val r = right.compute()
        val l = left.join()
        r.merge(l)
        return r
    }

    class ColorBins {
        var red = IntArray(COLOR_DEPTH)
        var green = IntArray(COLOR_DEPTH)
        var blue = IntArray(COLOR_DEPTH)

        internal var maxRed: Int = 0
        internal var maxGreen: Int = 0
        internal var maxBlue: Int = 0

        val maxFreq: Int
            get() = Math.max(Math.max(maxRed, maxGreen), maxBlue)

        internal fun merge(b: ColorBins) {
            for (i in 0 until COLOR_DEPTH) {
                red[i] += b.red[i]
                green[i] += b.green[i]
                blue[i] += b.blue[i]

                maxRed = Math.max(red[i], maxRed)
                maxGreen = Math.max(green[i], maxGreen)
                maxBlue = Math.max(blue[i], maxBlue)
            }
        }
    }

    companion object {
        private val TAG = Histogram::class.java.simpleName
        private const val sThreshold = 10000
        private const val COLOR_DEPTH = 256

        fun createHistogram(bmp: Bitmap): Histogram {
            val pixels = IntArray(bmp.width * bmp.height)
            bmp.getPixels(pixels, 0, bmp.width, 0, 0, bmp.width, bmp.height)
            return Histogram(pixels, 0, pixels.size)
        }
    }
}
